Start working

Katze719 opened the pull request.

Katze719 committed:
Add serial communication functionality with CMake configuration

- Introduced new source files for serial communication: `serial_open.cpp`, `serial_close.cpp`, `serial_read.cpp`, and `serial_write.cpp`.
- Implemented error handling and callback mechanisms for serial operations.
- Added a test suite for serial communication with Arduino, covering open/close, read/write operations, and timeout behavior.
- Configured CMake to copy compile commands to the project root for easier debugging.
- Removed the placeholder source file to clean up the project structure.

Katze719 committed:
Refactor error handling in serial communication files to use std::error_code for improved error messages

- Replaced usage of strerror(errno) with std::error_code for generating error messages in `serial_open.cpp`, `serial_close.cpp`, `serial_read.cpp`, and `serial_write.cpp`.
- Updated includes to replace <cstring> with <system_error> for better error handling practices.

Katze719 committed:
Add integration tests and Deno configuration for cpp-bindings-linux

- Created a new `.gitignore` file to exclude Deno artifacts and test snapshots.
- Added `deno.json` for task configuration and dependency management.
- Introduced `deno.lock` to lock dependencies for consistent builds.
- Implemented `ffi_bindings.ts` for FFI bindings to the cpp-bindings-linux library.
- Developed `integration_test.ts` to cover various scenarios including error handling, invalid parameters, and real serial port interactions.

Katze719 committed:
Refactor integration tests for improved clarity and error handling

- Updated error callback comments in `ffi_bindings.ts` for better readability.
- Reorganized status codes in `integration_test.ts` to use negative values for error handling.
- Simplified assertions in tests to directly compare expected and actual values, enhancing test clarity.
- Removed unnecessary comments and code to streamline the test suite.

Katze719 committed:
Refactor error callback in `ffi_bindings.ts` for improved clarity

- Simplified the `createErrorCallback` function by removing unnecessary destructuring of arguments.
- Enhanced type definitions for better TypeScript support and readability.

Katze719 committed:
Enhance Deno configuration and refactor FFI bindings for improved readability

- Added formatting options in `deno.json` for consistent code style.
- Refactored `ffi_bindings.ts` and `integration_test.ts` to improve code readability by standardizing indentation and line width.
- Updated integration tests to maintain clarity and consistency in test structure.

Katze719 committed:
Delete LICENSE

Mqxx commented on LICENSE:
Why removed? Should be replaced with [LGPL-2.0](https://www.gnu.org/licenses/old-licenses/lgpl-2.0.txt)

Mqxx commented on integration_tests/.gitignore:
```diff
@@ -0,0 +1,6 @@
+# Deno
+.deno/
```
For what is this directory?

Mqxx commented on integration_tests/deno.json:
```diff
@@ -0,0 +1,12 @@
+  },
+  "fmt": {
+    "lineWidth": 120,
+    "indentWidth": 4
```
Should be 2 but okay. 2 spaces are common for JS/TS.

Katze719 replied:
then why the comment, if okay

Mqxx replied:
To remind you. You want your formatting, I want mine. Because in the end you are writing this TS code and if you can read files better with 4 spaces than fine.

Mqxx commented on integration_tests/ffi_bindings.ts:
```diff
@@ -0,0 +1,166 @@
+        multiplier: number,
+        errorCallback: Deno.PointerValue | null,
+    ) => number;
+};
```
`Deno.PointerValue` should already be a union type containing `null`

```suggestion
export type SerialLib = {
    serialOpen: (
        port: Deno.PointerValue,
        baudrate: number,
        dataBits: number,
        parity: number,
        stopBits: number,
        errorCallback: Deno.PointerValue,
    ) => bigint;

    serialClose: (
        handle: bigint,
        errorCallback: Deno.PointerValue,
    ) => number;

    serialRead: (
        handle: bigint,
        buffer: Deno.PointerValue,
        bufferSize: number,
        timeoutMs: number,
        multiplier: number,
        errorCallback: Deno.PointerValue,
    ) => number;

    serialWrite: (
        handle: bigint,
        buffer: Deno.PointerValue,
        bufferSize: number,
        timeoutMs: number,
        multiplier: number,
        errorCallback: Deno.PointerValue,
    ) => number;
} as const;
```

Mqxx replied:
Resolved but did not change?

Katze719 replied:
null is still valid and could be the type no? for errorCallback

Katze719 replied:
not valid syntax: `Cannot find name 'as'.deno-ts(2304)`

Mqxx commented on integration_tests/ffi_bindings.ts:
```diff
@@ -0,0 +1,166 @@
+                symbols: SerialLib;
+                close: () => void;
+            };
+            lib = loaded;
```
```suggestion
            lib = Deno.dlopen(path, symbols);
            
```

Mqxx replied:
Resolved but did not change?

Katze719 replied:
i like it a bit more explicit, maybe i will have todo some clean up later

Mqxx commented on integration_tests/ffi_bindings.ts:
```diff
@@ -0,0 +1,166 @@
+    const definition = {
+        parameters: ["i32", "pointer"] as const,
+        result: "void" as const,
+    } as const;
```
```suggestion
    const definition = {
        parameters: ["i32", "pointer"],
        result: "void"
    } as const;
```

Mqxx commented on integration_tests/integration_test.ts:
All of these functions shouldn't be part of your testing scope. I will implement these tests on my side.

Mqxx commented on integration_tests/integration_test.ts:
I would suggest, refactoring the integration test cases to separate files. Each test should only test one behavior. Currently you have all test cases in one file, this is bad for maintainability. Please split the files to the corresponding test cases:
- `serial/open.test.ts`
- `serial/close.test.ts`
- `serial/read.test.ts`
- `serial/write.test.ts`

Also maybe remove these tests entirely and only test if you are able to load the library and register all the symbols. Testing individual functions is my task in the TypeScript library.

Mqxx commented on integration_tests/integration_test.ts:
```diff
@@ -0,0 +1,359 @@
+    },
+    sanitizeResources: false,
+    sanitizeOps: false,
+});
```
Also this file is way to long. Even for an integration test.

Mqxx commented on src/serial_close.cpp:
```diff
@@ -0,0 +1,32 @@
+extern "C"
+{
+
+    MODULE_API auto serialClose(int64_t handle, ErrorCallbackT error_callback) -> int
```
Is the documentation for these functions in the header file? Just a general question.

Katze719 replied:
should be, they are implemented as defined in cpp_core

Mqxx commented on src/serial_open.cpp:
```diff
@@ -0,0 +1,201 @@
+
+extern "C"
+{
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity)
```
This is a linting rule right? Any it's complaining because there are to many arguments for this function? If so, instead of using the crowbar to disable the linter here, please change the configuration for the linter instead to accept more arguments in the function head.

Katze719 replied:
it says the function is to complex and should be refactored, to many if statements, to long, etc. is WIP

Mqxx replied:
Ahh okay. Yea I see that... Maybe refactor to helper functions to simplify logic.

Mqxx commented on src/serial_open.cpp:
```diff
@@ -0,0 +1,201 @@
+
+        const char *port_path = static_cast<const char *>(port);
+
+        const int fd = open(port_path, O_RDWR | O_NOCTTY | O_NONBLOCK);
```
I am assuming that `fd` stands for `file_descriptor`? Pleas use variable names that are clearly.

```suggestion
        const int handle = open(port_path, O_RDWR | O_NOCTTY | O_NONBLOCK);
```

Katze719 replied:
fd is the kernel default for file descriptors, since its not just a handle

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+#include <system_error>
+#include <unistd.h>
+
+// NOLINTNEXTLINE(misc-use-anonymous-namespace)
```
Is this needed?

Katze719 replied:
WIP

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+
+extern "C"
+{
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity)
```
Same here with the linter rule as described above.

Katze719 replied:
same as above

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+extern "C"
+{
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity)
+    MODULE_API auto serialRead(int64_t handle, void *buffer, int buffer_size, int timeout_ms, int /*multiplier*/,
```
Why is the `multiplier` commented out?

Katze719 replied:
WIP

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+            return static_cast<int>(cpp_core::StatusCodes::kInvalidHandleError);
+        }
+
+        const int fd = static_cast<int>(handle);
```
Same here with `fd`.

Katze719 replied:
same with fd here

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+        while (elapsed_ms < timeout_ms)
+        {
+            struct pollfd pfd = {};
+            pfd.fd = fd;
```
`pfd`? Please change accordingly.

Katze719 replied:
the struct is called pollfd, and is one line above

Mqxx replied:
> the struct is called pollfd, and is one line above

And?

Katze719 replied:
what do you want me to call it? `pollfd` ? that would shadow the struct

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,158 @@
+        const int poll_interval = 10;
+        bool data_ready = false;
+
+        while (elapsed_ms < timeout_ms)
```
What if I want to read infinitely?

Katze719 replied:
infinitely waiting is not defined in https://github.com/Serial-IO/cpp-core/blob/main/include/cpp_core/interface/serial_read.h from cpp_core, and would result in permanently blocking and never returning. would result in hanging the programm until you kill the process? i dont think this is a valid use case

Mqxx replied:
We haven't implemented an abort controller yet, aka. a function that aborts the current read/write action. This could be used to infinitely read without quitting the program. [Here is an example on what an abort controller is in JS.](https://developer.mozilla.org/de/docs/Web/API/AbortController)

For you it would be just a "simple" `seriel_read_abort` function, that cancels the current read/write operation.

Katze719 replied:
after reading some documentation, reading infinitely is supported at the moment, just not documented, see 

https://www.man7.org/linux/man-pages/man2/poll.2.html

```
       Note that the timeout interval will be rounded up to the system
       clock granularity, and kernel scheduling delays mean that the
       blocking interval may overrun by a small amount.  Specifying a
       negative value in timeout means an infinite timeout.  Specifying a
       timeout of zero causes poll() to return immediately, even if no
       file descriptors are ready.
```

Katze719 replied:
since it still supports the full feature set specified in core_cpp, is see no problem here

Katze719 replied:
seriel_read_abort will do nothing if a timeout of infinite is specified, since the call is already made to the kernel

Mqxx replied:
Is there no way to cancel it?

Katze719 replied:
maybe with a pipe, but i think then the function would only return early and the kernel would still wait for a byte

Mqxx replied:
Okay. But doesn't polling mean that it will, well poll? aka loop? So is there not a way of checking before the next loop for an abort?

Katze719 replied:
yes, and for the abort it needs a timeout or a byte, or if the timeout is zero, then it will return instantly. what do you mean by checking? check if there is a byte?

Katze719 replied:
maybe creating a pipe and connect the pipe to the usb device and the other end to the poll() function, could resolve this via deleting the pipe/fd, will have to try that

Mqxx replied:
No. I thought the function works like this:

1. Check if bytes are present
2. Read amount of present bytes before timeout runs out
3. GoTo 1.

So I thought adding the following would add the abort controller:

0. Check if abort flag is set to true, exit if so
1. Check if bytes are present
2. Read amount of present bytes before timeout runs out
3. GoTo 0.

Katze719 replied:
maybe this could be resolved in another pull request? so this one keeps beeing small?

Katze719 replied:
you mean building a wrapper for poll, where you just loop on calling the real poll function, but what timeout do give the real poll function then? this could hurt performance

Katze719 replied:
poll works more like this:

1. call poll
2. poll checkfs if the fd is ready until timeout
3. is fd ready? return a nonzero number, did a timeout occur return 0, did a error happen return -1


 call read to read bytes from it

Mqxx replied:
Okay. Maybe implement this in another PR.

Katze719 replied:
pls open a issue for this feature

Mqxx commented on src/serial_write.cpp:
```diff
@@ -0,0 +1,97 @@
+#include <termios.h>
+#include <unistd.h>
+
+// NOLINTNEXTLINE(misc-use-anonymous-namespace)
```
Is this needed?

Katze719 replied:
WIP

Mqxx commented on src/serial_write.cpp:
```diff
@@ -0,0 +1,97 @@
+extern "C"
+{
+
+    MODULE_API auto serialWrite(int64_t handle, const void *buffer, int buffer_size, int timeout_ms, int /*multiplier*/,
```
Why is the `multiplier` commented out?

Katze719 replied:
WIP

Mqxx commented on tests/test_serial_arduino.cpp:
If possible use a tty emulator for this instead of physical hardware. That way you could also automate the test using CI.

Mqxx commented on integration_tests/integration_test.ts:
```diff
@@ -0,0 +1,359 @@
+
+        const result = lib.serialClose(BigInt(-1), null);
+        assertEquals(result, StatusCodes.kSuccess);
+        console.log("Invalid handle close handled correctly");
```
All these logs can be removed. If the test fails, we know and also if the test succeeds we know. No need for extra logging.

Katze719 committed:
Add CI workflow for C++ and Deno integration tests

- Introduced a GitHub Actions CI workflow in `ci.yml` to automate C++ and Deno integration tests.
- Configured separate jobs for C++ tests using GCC and Deno tests with specified versions.
- Set up environment variables and build dependencies for both test environments.
- Updated Deno integration tests to reflect minimal FFI bindings and removed unnecessary error handling functions.
- Simplified integration test structure to focus on library loading and unloading.

Katze719 committed:
Update CI workflow to use Fedora container and simplify GCC configuration

- Changed CI workflow to use a Fedora container for better CMake support.
- Removed matrix strategy for GCC versions, simplifying the job configuration.
- Updated package installation commands to use `dnf` instead of `apt-get` for Fedora compatibility.
- Adjusted environment variable settings to reflect the new container setup.

Katze719 committed:
Update CI workflow to install g++ and streamline CMake configuration

- Added g++ to the list of build dependencies in the CI workflow.
- Removed unnecessary comments related to GCC configuration for clarity.
- Ensured environment variable settings for CMake are properly configured.

Katze719 committed:
Add missing BOTHER definition in serial_open.cpp for compatibility with minimal environments

- Defined BOTHER in serial_open.cpp if not already defined, ensuring compatibility with certain libcs or kernel headers that may not include it.
- This change supports builds in minimal environments, such as Deno's Debian-based CI containers.

Mqxx commented on .github/workflows/deno-tests.yml:
We are not on GitLab. Don't be shy. You can also create multiple workflow files xD. Please refactor these to their own `cpp_test.yml` and `ts_test.yml` or at least rename tha file to `test.yml` instead of `ci.yml`

Katze719 committed:
Add CI workflows for building Deno shared library and running C++ tests

- Introduced a GitHub Actions workflow for building a shared library for Deno, including steps for dependency installation, CMake configuration, and artifact collection.
- Added a separate workflow for running C++ unit and integration tests, utilizing a Fedora container and configuring the environment for serial communication testing.
- Both workflows are triggered on pushes and pull requests to the main branches, ensuring automated testing and builds.

Katze719 committed:
Refactor serial communication error handling and introduce posix_helpers

- Replaced direct error handling in serial_open, serial_close, serial_read, and serial_write with utility functions from the new posix_helpers.hpp for improved code clarity and consistency.
- Introduced UniqueFd class to manage file descriptors safely and prevent resource leaks.
- Simplified error callback invocations to streamline error reporting across serial communication functions.

Katze719 committed:
Refactor serial_read to improve readability and modularity

- Moved the waitFdReady function into an anonymous namespace for better encapsulation.
- Simplified the serialRead function by introducing a lambda for reading data, enhancing clarity.
- Streamlined the logic for checking data readiness and reading from the file descriptor, reducing complexity and improving maintainability.

Katze719 committed:
Refactor serial communication functions to enhance clarity and maintainability

- Moved the waitFdReady function to posix_helpers.hpp for better code organization and reuse across serial_read and serial_write.
- Updated serial_read and serial_write implementations to utilize the new waitFdReady function, improving readability and reducing code duplication.
- Added detailed comments for parity and stop_bits configurations in serial_open.cpp to clarify their usage and constraints.

Katze719 committed:
Update integration_tests/ffi_bindings.ts

Co-authored-by: Mqx <62719703+Mqxx@users.noreply.github.com>

Katze719 committed:
Update integration_tests/ffi_bindings.ts to make serial function signatures readonly

- Changed function signatures in the SerialLib type to readonly, ensuring immutability and enhancing type safety.
- This update improves the clarity of the API by indicating that these functions should not be modified after their initial definition.

Katze719 committed:
Update CI workflow to use latest Ubuntu and optimize dependency installation

- Changed the runner from ubuntu-20.04 to ubuntu-latest for improved compatibility and support.
- Added a container specification to use Ubuntu 20.04.
- Updated the dependency installation process to use non-interactive mode and removed unnecessary package lists to streamline the build process.

Katze719 committed:
Update CI workflow to use Ubuntu 24.04 container for improved compatibility

Katze719 committed:
Enhance handle validation in serial communication functions

- Added checks in serial_close, serial_read, and serial_write to ensure that handles do not exceed the maximum limit for integers, improving error handling and robustness.
- Included <limits> header to facilitate the validation of handle values across the functions.

Katze719 committed:
Refactor FFI bindings in integration_tests/ffi_bindings.ts for improved type safety

- Updated the SerialLib type to utilize Deno.DynamicLibrary for better type inference and clarity.
- Consolidated symbol definitions into a single object, enhancing organization and maintainability.
- Adjusted the loadSerialLib function to reflect the new type structure, ensuring consistency across the codebase.

Katze719 committed:
Add JSR package support for Linux shared library

- Introduced a new GitHub Actions workflow for publishing the native Linux shared library as a JSON/base64 blob to JSR.
- Added scripts for embedding the shared library and generating the corresponding JSON metadata.
- Created TypeScript bindings for the shared library, enhancing usability and type safety.
- Updated CMake configuration to generate JSR package metadata during the build process.
- Included a README for usage instructions and permissions required for the library.

Katze719 committed:
Update CMakeLists.txt to bump version tags for dependencies

- Updated cmake_git_versioning GIT_TAG from v1.0.0 to v1.0.1 for the latest features and fixes.
- Changed cpp_core GIT_TAG from main to bump-version-from-git-versioning to align with versioning strategy.

Katze719 committed:
Add permissions to GitHub Actions workflow for JSR publishing

- Updated the publish-jsr.yml workflow to include permissions for reading contents and writing ID tokens, enhancing security and functionality during the publishing process.

Katze719 committed:
Add LGPL-3.0-only license and update package metadata

- Introduced a new LICENSE file containing the GNU Lesser General Public License v3.0.
- Updated jsr.json.in to specify the license as LGPL-3.0-only.
- Added a license section in the README.md to inform users about the licensing terms.

Katze719 committed:
Update JSR publishing workflow to allow dirty worktree during package publishing

- Modified the publish-jsr.yml workflow to include the --allow-dirty flag for deno publish, enabling the publishing process to proceed even when the worktree is dirty.
- Added comments to clarify the impact of CMake and embed_so.ts on the worktree state, ensuring deterministic builds and generation in a single run.

Katze719 committed:
Refactor JSR package structure and remove FFI bindings

- Updated the JSR package to use a new directory structure for binaries, changing the path from `jsr/binaries/linux-x86_64.json` to `jsr/bin/linux-x86_64.json`.
- Deleted the `ffi.ts` and `mod.ts` files, removing the TypeScript bindings for the shared library.
- Updated the README.md to reflect the new usage instructions for importing the JSON and writing the shared library to disk.
- Added a new `linux-x86_64.json` file to store the shared library metadata.

Katze719 committed:
Update JSR package to use new x84_64 naming convention for binaries

- Changed references from `linux-x86_64` to `x84_64` in the JSR package configuration and scripts.
- Updated the JSON metadata file to reflect the new naming and added the new `x84_64.json` file.
- Modified the README.md and embed_so.ts script to align with the new import path for the shared library.

Katze719 committed:
Remove exclusion of scripts from JSR package configuration in jsr.json.in

Mqxx commented on jsr/README.md:
```diff
@@ -0,0 +1,47 @@
+## License
+
+This package is licensed under **LGPL-3.0-only** (see the repository root
+`LICENSE`).
```
```suggestion
# C++ Bindings Linux

This package ships the native Linux shared library payload as a **JSON/base64 blob**.

## Usage

Import the JSON and write the `.so` to disk (consumer project example):

```ts
import blob from "@serial/cpp-bindings-linux/x84_64" with {type: 'json'};

const bytes = new TextEncoder().encode(atob(blob.data));

const tempFilePath = Deno.makeTempFileSync();
Deno.writeFileSync(tempFilePath, bytes, { mode: 0o755 });

// Now you can open the binary using for example `Deno.dlopen`
```

## License

This package is licensed under **LGPL-3.0-only** (see the repository root `LICENSE`).
```

Katze719 committed:
Update jsr/README.md

Co-authored-by: Mqx <62719703+Mqxx@users.noreply.github.com>

Katze719 committed:
Update README.md to improve formatting and add license section

- Adjusted formatting of the JSON/base64 blob description for clarity.
- Added a new section to specify the package license as LGPL-3.0-only.

Katze719 committed:
Update CMakeLists.txt to change cpp_core GIT_TAG from bump-version-from-git-versioning to main
